File: src/main/kotlin/com/babasama/edendale/domain/MediaModels.kt
package com.babasama.edendale.domain

/** The two title types returned by TMDB browse and search endpoints. */
enum class MediaType(val pathSegment: String) {
    MOVIE("movie"),
    TV("tv"),
}

/** Lightweight, platform-neutral handle used to request a title's details. */
data class MediaRef(
    val id: Int,
    val mediaType: MediaType,
)

/** A movie or show rendered in shelves, search results, and collections. */
data class MediaItem(
    val id: Int,
    val mediaType: MediaType,
    val title: String,
    val overview: String? = null,
    val posterPath: String? = null,
    val backdropPath: String? = null,
    val voteAverage: Double? = null,
    val releaseDate: String? = null,
) {
    val ref: MediaRef get() = MediaRef(id, mediaType)
    val year: Int?
        get() = releaseDate
            ?.takeIf { it.length >= 4 }
            ?.take(4)
            ?.toIntOrNull()

    fun posterUrl(size: TmdbImageSize = TmdbImageSize.POSTER): String? =
        tmdbImageUrl(posterPath, size)

    fun backdropUrl(size: TmdbImageSize = TmdbImageSize.BACKDROP): String? =
        tmdbImageUrl(backdropPath, size)
}

enum class TmdbImageSize(val pathSegment: String) {
    POSTER("w342"),
    POSTER_LARGE("w500"),
    BACKDROP("w780"),
    /** Cast-row and people-grid portraits. */
    PROFILE("w185"),
    /** The person page's hero portrait. */
    PROFILE_LARGE("h632"),
    ORIGINAL("original"),
}

fun tmdbImageUrl(path: String?, size: TmdbImageSize): String? =
    path?.let { "https://image.tmdb.org/t/p/${size.pathSegment}$it" }

data class Genre(
    val id: Int,
    val name: String,
)

/** An actor/actress from people search: portrait plus best-known titles. */
data class PersonItem(
    val id: Int,
    val name: String,
    val profilePath: String? = null,
    val knownFor: List<String> = emptyList(),
) {
    fun profileUrl(size: TmdbImageSize = TmdbImageSize.PROFILE): String? =
        tmdbImageUrl(profilePath, size)
}

data class CastMember(
    val id: Int,
    val name: String,
    val character: String? = null,
    val profilePath: String? = null,
) {
    fun profileUrl(size: TmdbImageSize = TmdbImageSize.PROFILE): String? =
        tmdbImageUrl(profilePath, size)
}

/**
 * A person's full TMDB record — the header of a person page. The filmography
 * itself stays a separate call (`/person/{id}/combined_credits`) so the two
 * can be fetched concurrently.
 */
data class PersonDetail(
    val id: Int,
    val name: String,
    val biography: String? = null,
    val profilePath: String? = null,
    val birthday: String? = null,
    val deathday: String? = null,
    val placeOfBirth: String? = null,
    val knownForDepartment: String? = null,
) {
    fun profileUrl(size: TmdbImageSize = TmdbImageSize.PROFILE_LARGE): String? =
        tmdbImageUrl(profilePath, size)

    /**
     * "1956 – 2016 · Concord, California" — the vitals line under the name.
     * Null when TMDB knows neither a date nor a place.
     */
    val vitals: String?
        get() {
            val born = birthday?.take(4)?.takeIf { it.length == 4 }
            val died = deathday?.take(4)?.takeIf { it.length == 4 }
            val lifespan = when {
                born != null && died != null -> "$born – $died"
                born != null -> born
                died != null -> "– $died"
                else -> null
            }
            return listOfNotNull(lifespan, placeOfBirth)
                .takeIf { it.isNotEmpty() }
                ?.joinToString(" · ")
        }
}

/** One entry in a show's season list, surfaced on [MediaDetail.seasons]. */
data class SeasonSummary(
    val seasonNumber: Int,
    val name: String,
    val episodeCount: Int? = null,
    val airDate: String? = null,
    val posterPath: String? = null,
    val overview: String? = null,
) {
    fun posterUrl(size: TmdbImageSize = TmdbImageSize.POSTER): String? =
        tmdbImageUrl(posterPath, size)
}

/** Full title data used by Android's hero and detail UIs. */
data class MediaDetail(
    val ref: MediaRef,
    val title: String,
    val tagline: String? = null,
    val overview: String? = null,
    val posterPath: String? = null,
    val backdropPath: String? = null,
    val year: Int? = null,
    val runtimeMinutes: Int? = null,
    val genres: List<String> = emptyList(),
    val attribution: String? = null,
    val score: Double? = null,
    val voteCount: Int? = null,
    val cast: List<CastMember> = emptyList(),
    val seasonCount: Int? = null,
    val episodeCount: Int? = null,
    val seasons: List<SeasonSummary> = emptyList(),
) {
    fun posterUrl(size: TmdbImageSize = TmdbImageSize.POSTER_LARGE): String? =
        tmdbImageUrl(posterPath, size)

    fun backdropUrl(size: TmdbImageSize = TmdbImageSize.ORIGINAL): String? =
        tmdbImageUrl(backdropPath, size)
}

enum class WatchMediaType { MOVIE, EPISODE }

/** Persisted Android watch state. */
data class WatchProgress(
    val tmdbId: Int,
    val mediaType: WatchMediaType,
    val position: Double = 0.0,
    val watchedSeconds: Double = 0.0,
    val lastWatchedEpochMillis: Long = 0L,
    val isCompleted: Boolean = false,
    val showTmdbId: Int? = null,
    val seasonNumber: Int? = null,
    val episodeNumber: Int? = null,
) {
    val normalizedPosition: Double get() = position.coerceIn(0.0, 1.0)
    val storageKey: String get() = "${mediaType.name.lowercase()}:$tmdbId"
}

sealed interface ParsedMedia {
    data class Movie(val title: String, val year: Int?) : ParsedMedia
    data class Episode(val showName: String, val season: Int, val episode: Int) : ParsedMedia
}

File: Edendale.Windows/Models/MediaModels.cs
// Windows media and TMDB domain models.

using System.Text.Json.Serialization;

namespace Edendale.Windows.Models;

public sealed class MediaRef
{
    public int Id { get; set; }
    public string MediaType { get; set; } = "movie";

    public bool IsShow => MediaType == "tv";
}

public sealed class MediaItem
{
    public int Id { get; set; }
    public string MediaType { get; set; } = "movie";
    public string Title { get; set; } = "";
    public string? Overview { get; set; }
    public string? PosterPath { get; set; }
    public string? BackdropPath { get; set; }
    public string? PosterUrl { get; set; }
    public string? BackdropUrl { get; set; }
    public double? VoteAverage { get; set; }
    public string? ReleaseDate { get; set; }
    public int? Year { get; set; }

    public MediaRef Ref => new() { Id = Id, MediaType = MediaType };
}

public sealed class Genre
{
    public int Id { get; set; }
    public string Name { get; set; } = "";
}

/// <summary>One day cell of the release heatmap; null slots pad the year's edges.</summary>
public sealed class ReleaseDaySlot
{
    public string DateKey { get; set; } = "";
    public int Year { get; set; }
    public int Month { get; set; }
    public int Day { get; set; }
    public bool IsFirstOfMonth { get; set; }
}

/// <summary>One week of the heatmap: seven slots, Sunday first.</summary>
public sealed class ReleaseWeekColumn
{
    public string? MonthLabel { get; set; }
    public List<ReleaseDaySlot?> Slots { get; set; } = [];
}

/// <summary>A year unrolled into week columns — the heatmap's layout.</summary>
public sealed class ReleaseYearGrid
{
    public int Year { get; set; }
    public List<ReleaseWeekColumn> Columns { get; set; } = [];
}

public sealed class CastMember
{
    public int Id { get; set; }
    public string Name { get; set; } = "";
    public string? Character { get; set; }
    public string? ProfilePath { get; set; }
    public string? ProfileUrl { get; set; }
}

/// <summary>An actor/actress row from people search.</summary>
public sealed class PersonItem
{
    public int Id { get; set; }
    public string Name { get; set; } = "";
    public string? ProfilePath { get; set; }
    public string? ProfileUrl { get; set; }
    public List<string> KnownFor { get; set; } = [];

    public string KnownForText => string.Join(" · ", KnownFor);
}

/// <summary>
/// /person/{id} — the biography header of a person page. The filmography is
/// a separate call so the two can be fetched concurrently.
/// </summary>
public sealed class PersonDetail
{
    public int Id { get; set; }
    public string Name { get; set; } = "";
    public string? Biography { get; set; }
    public string? ProfilePath { get; set; }
    public string? ProfileUrl { get; set; }
    public string? Birthday { get; set; }
    public string? Deathday { get; set; }
    public string? PlaceOfBirth { get; set; }
    public string? KnownForDepartment { get; set; }

    /// <summary>"1956 – 2016 · Concord, California"; empty when unknown.</summary>
    public string? Vitals { get; set; }

    public bool HasBiography => !string.IsNullOrWhiteSpace(Biography);
    public bool HasVitals => !string.IsNullOrWhiteSpace(Vitals);
    public bool HasDepartment => !string.IsNullOrWhiteSpace(KnownForDepartment);
}

/// <summary>
/// One scoped search: titles and people, plus the scope its keyword prefix
/// asked for. <see cref="LeadsWithPeople"/> decides which section renders first.
/// </summary>
public sealed class ScopedSearchResult
{
    /// <summary>all | people | movies | shows.</summary>
    public string Scope { get; set; } = "all";

    /// <summary>The query with any keyword prefix stripped.</summary>
    public string Term { get; set; } = "";

    public bool LeadsWithPeople { get; set; }
    public List<MediaItem> Titles { get; set; } = [];
    public List<PersonItem> People { get; set; } = [];

    public bool IsEmpty => Titles.Count == 0 && People.Count == 0;
}

public sealed class MediaDetail
{
    public MediaRef Ref { get; set; } = new();
    public string Title { get; set; } = "";
    public string? Tagline { get; set; }
    public string? Overview { get; set; }
    public string? PosterPath { get; set; }
    public string? BackdropPath { get; set; }
    public string? PosterUrl { get; set; }
    public string? BackdropUrl { get; set; }
    public int? Year { get; set; }
    public int? RuntimeMinutes { get; set; }
    public List<string> Genres { get; set; } = [];
    public string? Attribution { get; set; }
    public double? Score { get; set; }
    public int? VoteCount { get; set; }
    public List<CastMember> Cast { get; set; } = [];
    public int? SeasonCount { get; set; }
    public int? EpisodeCount { get; set; }

    /// <summary>
    /// Seasons in picker order — numbered ascending, Specials last, empty
    /// seasons already dropped by the Windows TMDB mapping.
    /// </summary>
    public List<SeasonSummary> Seasons { get; set; } = [];
}

public sealed class SeasonSummary
{
    public int SeasonNumber { get; set; }
    public string Name { get; set; } = "";
    public int? EpisodeCount { get; set; }
    public string? AirDate { get; set; }
    public string? PosterPath { get; set; }
    public string? PosterUrl { get; set; }
    public string? Overview { get; set; }

    /// <summary>"Season 2" for numbered seasons, TMDB's own name for Specials.</summary>
    public string DisplayTitle => SeasonNumber == 0 ? Name : $"Season {SeasonNumber}";
}

public sealed class SeasonDetail
{
    public int SeasonNumber { get; set; }
    public string Name { get; set; } = "";
    public string? Overview { get; set; }
    public string? AirDate { get; set; }
    public string? PosterPath { get; set; }
    public string? PosterUrl { get; set; }
    public List<EpisodeDetail> Episodes { get; set; } = [];
}

public sealed class WatchProgress
{
    public int TmdbId { get; set; }
    /// <summary>"movie" or "episode".</summary>
    public string MediaType { get; set; } = "movie";
    public double Position { get; set; }
    public double NormalizedPosition { get; set; }
    public double WatchedSeconds { get; set; }
    public long LastWatchedEpochMillis { get; set; }
    public bool IsCompleted { get; set; }
    public int? ShowTmdbId { get; set; }
    public int? SeasonNumber { get; set; }
    public int? EpisodeNumber { get; set; }
    public string StorageKey => $"{MediaType}:{TmdbId}";
}

public sealed class HeroScene
{
    public MediaDetail Detail { get; set; } = new();
    public WatchProgress? Progress { get; set; }
    public bool IsContinueWatching { get; set; }
    public string? RemainingText { get; set; }
}

public sealed class CollectionFilter
{
    public string Id { get; set; } = "all";
    public string Title { get; set; } = "";
    public int? GenreId { get; set; }
}

public sealed class HomeCatalog
{
    public List<HeroScene> HeroScenes { get; set; } = [];
    public List<MediaItem> Trending { get; set; } = [];
    public List<MediaItem> PopularMovies { get; set; } = [];
    public List<MediaItem> PopularShows { get; set; } = [];
    public List<MediaItem> TopRated { get; set; } = [];
    public List<Genre> Genres { get; set; } = [];
    public List<CollectionFilter> Collections { get; set; } = [];
}

public sealed class EpisodeDetail
{
    public int Id { get; set; }
    public string Name { get; set; } = "";
    public string? Overview { get; set; }
    public string? StillPath { get; set; }
    public string? StillUrl { get; set; }
    public string? AirDate { get; set; }
    public int? RuntimeMinutes { get; set; }
    public int? SeasonNumber { get; set; }
    public int? EpisodeNumber { get; set; }
    public double? VoteAverage { get; set; }
}

public sealed class TrailerVideo
{
    public string Id { get; set; } = "";
    public string Key { get; set; } = "";
    public string? Name { get; set; }
    public string? Site { get; set; }
    public string? Type { get; set; }
    public bool? Official { get; set; }
}

/// <summary>Result of local filename classification: "movie" or "episode".</summary>
public sealed class ParsedMedia
{
    public string Kind { get; set; } = "movie";
    public string? Title { get; set; }
    public int? Year { get; set; }
    public string? ShowName { get; set; }
    public int? Season { get; set; }
    public int? Episode { get; set; }

    [JsonIgnore]
    public bool IsEpisode => Kind == "episode";
}
